### ISSUE TITLE
Oops! Database Error: UNIQUE constraint failed. Library is not corrupt

### ISSUE THREAD

**Issue by kanjieater** [janeczku/calibre-web#3058]
**Describe the bug/problem** 

Books uploaded to Calibre-web fails. Book added to classic calibre succeeds. Ran Library Maintenance, fix and deleted, found no issues, issue persists. Oops! Database Error: UNIQUE constraint failed: publishers.name.

Publisher in question: 
株式会社 東京創元社

![image](https://github.com/janeczku/calibre-web/assets/32607317/11f0dc26-c182-420d-a456-51377d8ab3e2)

**To Reproduce**

1. Drag and drop [庵野 ゆき] 竜の医師団１ (創元推理文庫).epub into library
2. Error generated and fails to add book
Not that other books add fine.
**Logfile**

--- Logging error ---
```
Traceback (most recent call last):
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlite3.IntegrityError: UNIQUE constraint failed: publishers.name

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/app/calibre-web/cps/editbooks.py", line 257, in upload
    db_book, input_authors, title_dir, renamed_authors = create_book_on_upload(modify_date, meta)
  File "/app/calibre-web/cps/editbooks.py", line 712, in create_book_on_upload
    calibre_db.session.flush()
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3449, in flush
    self._flush(objects)
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3588, in _flush
    with util.safe_reraise():
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/util/langhelpers.py", line 70, in __exit__
    compat.raise_(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/session.py", line 3549, in _flush
    flush_context.execute()
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 456, in execute
    rec.execute(self)
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/unitofwork.py", line 630, in execute
    util.preloaded.orm_persistence.save_obj(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 237, in save_obj
    _emit_update_statements(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/orm/persistence.py", line 1001, in _emit_update_statements
    c = connection._execute_20(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1710, in _execute_20
    return meth(self, args_10style, kwargs_10style, execution_options)
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/sql/elements.py", line 334, in _execute_on_connection
    return connection._execute_clauseelement(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1577, in _execute_clauseelement
    ret = self._execute_context(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1953, in _execute_context
    self._handle_dbapi_exception(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 2134, in _handle_dbapi_exception
    util.raise_(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/util/compat.py", line 211, in raise_
    raise exception
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/base.py", line 1910, in _execute_context
    self.dialect.do_execute(
  File "/lsiopy/lib/python3.10/site-packages/sqlalchemy/engine/default.py", line 736, in do_execute
    cursor.execute(statement, parameters)
sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) UNIQUE constraint failed: publishers.name
[SQL: UPDATE publishers SET name=? WHERE publishers.id = ?]
[parameters: ('株式会社 東京創元社', 235)]
(Background on this error at: https://sqlalche.me/e/14/gkpj)
```

**Expected behavior**

Books that can be added to Calibre should be able to be added to Calibre-Web without having to open Calibre

**Environment (please complete the following information):**

 - OS: Windows
 - Python version: 3.10.10
 - Calibre-Web version: [e.g. 0.6.8 or 087c4c59 (git rev-parse --short HEAD)]:
 - Docker container: [None/LinuxServer]: LinuxServer
 - Special Hardware: Windows PC
 - Browser: Chrome 

**Additional context**
Add any other context about the problem here. [e.g. access via reverse proxy, database background sync, special database location]

**Comment by OzzieIsaacs** [comment]
There is a publish with the same name, but different in the database. I don't know how to give an example form JKC letters but for normal ascii characters it would be something like "Unknown" and "unknown" (Capital letter or not)

**Comment by kanjieater** [comment]
> There is a publish with the same name, but different in the database. I don't know how to give an example form JKC letters but for normal ascii characters it would be something like "Unknown" and "unknown" (Capital letter or not)

How do you avoid this? Only add books via calibre-web? Reset my library and let calibre-web handle all imports?

**Comment by xmgz** [comment]
@kanjieater in my particular case, with you same error (`publishers.name`  key value) I had to manually edit the books in my library with that particular publisher name that caused the error.

go to "book list" (last item in the left dashboar menu); order by "editor" and look for the publisher name that brings your error; **manually** edit  (click on its name) deleting its value in every book that has this publisher.  In my case there were only three.

then try to add a new book and it worked for me

then manually edit again those previous books adding back its publisher name.


I know, this is not how it should work, but at least I could add new books to library

**Comment by aserrallerios** [comment]
I'm having the same issue with certain book:

```
Oops! Database Error: UNIQUE constraint failed: tags.name.
``` 

I only add books from the calibre-web UI.

I run the calibre-web in docker in TrueNAS, and somehow the logs of the application are not shown, so the only alternative I got was to copy the `metadata.db` to a local calibre and test again.

The database loaded successfully and I ran a validation of the database that showed no errors. Then I tried to add the book and it was added with no problems. Now I can see the `tags` of the book:

```
Historia, Religion
```
 
<img width="276" alt="Screenshot 2024-06-14 at 15 27 07" src="https://github.com/janeczku/calibre-web/assets/3159593/dbd47718-ff85-4ad0-a617-9bf8fa6a86ce">

<img width="264" alt="Screenshot 2024-06-14 at 15 27 31" src="https://github.com/janeczku/calibre-web/assets/3159593/7b1db44c-ed93-4fb1-af91-ff410e2c2b09">

So basically the suggestion is to update the DB locally and upload it to calibre-web every time I get this type of error?

**Comment by kaibagley** [comment]
I am also having this issue, but for `authors.name`. I had a book called The Cyberiad by Stanislaw Lem, and was trying to upload Solaris from the same author.

I saw that The Cyberiad was under the author Stanisław Lem (with a ł), and the book I was uploading mustn't have had the ł character, because when I renamed the author in The Cyberiad to be Stanislaw, I was then able to upload Solaris.

Hope this helps!

**Comment by OzzieIsaacs** [comment]
@kaibagley: I can reproduce it

**Comment by OzzieIsaacs** [comment]
It's a bit tricky, I need some tests, the solution will show up within next week

**Comment by NicholasYZ** [comment]
For some additional information, the error message I encountered was `UNIQUE constraint failed: authors.name`.

The uploaded epub files' author name `理不盡な孫の手` was conflicted with the author name `理不尽な孙の手` of books already inside the database.The difference inbetween is Japanese kanji/traditional Chinese and simplified Chinese characters (its actually a Japanese name, but different translators chose different approaches when dealing with the author name).

I suspect that some kind of pluggin identify these two as the same key but with different characters, causing the unique contraint in database to fail. Therefore I changed the author name into `理不尽な孙の手` using calibre ebook editior, specifically speaking, the `<dc:creator opf:role="aut"> </dc:creator>` line in `content.opf` page, and it worked just fine. The book after editing could be uploaded and identified just like any other books.

Hope this would help, and thanks again for all your efforts put in this project.

**Comment by OzzieIsaacs** [comment]
I‘m still working on solving the issue, there are 2 problems left in conjunction with google drive

**Comment by OzzieIsaacs** [comment]
It's fixed and released in 0.6.22

**Comment by aserrallerios** [comment]
Testing  new image:

```
    Image:          linuxserver/calibre-web:version-0.6.22
    Image ID:       docker.io/linuxserver/calibre-web@sha256:2de4ab91f2768c6db32d083a5935663fcc0eded1a67a4fdc2c1f705f7bb24d2d
```

I got this error:

```
[2024-07-08 00:09:53,826] ERROR {cps:1414} Exception on /upload [POST]
Traceback (most recent call last):
  File "/lsiopy/lib/python3.10/site-packages/flask/app.py", line 2190, in wsgi_app
    response = self.full_dispatch_request()
  File "/lsiopy/lib/python3.10/site-packages/flask/app.py", line 1486, in full_dispatch_request
    rv = self.handle_user_exception(e)
  File "/lsiopy/lib/python3.10/site-packages/flask/app.py", line 1484, in full_dispatch_request
    rv = self.dispatch_request()
  File "/lsiopy/lib/python3.10/site-packages/flask/app.py", line 1469, in dispatch_request
    return self.ensure_sync(self.view_functions[rule.endpoint])(**view_args)
  File "/app/calibre-web/cps/usermanagement.py", line 34, in decorated_view
    return func(*args, **kwargs)
  File "/app/calibre-web/cps/editbooks.py", line 59, in inner
    return f(*args, **kwargs)
  File "/app/calibre-web/cps/editbooks.py", line 249, in upload
    meta, error = file_handling_on_upload(requested_file)
  File "/app/calibre-web/cps/editbooks.py", line 755, in file_handling_on_upload
    if not validate_mime_type(requested_file, allowed_extensions):
  File "/app/calibre-web/cps/file_helper.py", line 49, in validate_mime_type
    mime = magic.Magic(mime=True)
NameError: name 'magic' is not defined
```

It's probably unrelated, but I'm not sure what could be wrong.

**Comment by OzzieIsaacs** [comment]
Please install python-magic

**Comment by aserrallerios** [comment]
But this is the docker image. How I'm supposed to do that?

**Comment by OzzieIsaacs** [comment]
Please ask the docker provider

**Comment by aserrallerios** [comment]
@OzzieIsaacs I'm using the last docker image with magic installed, but I still get this error:

```
Oops! Database Error: UNIQUE constraint failed: tags.name.
```

Is it possible that you fixed it for other attributes but not for `tags.name`?

**Comment by OzzieIsaacs** [comment]
@aserrallerios For me it's working (condition: only books with e.g. tag name: Religion and no book with tag religion). Could you please rename the 4 books with "religion" to "Religion" and check again, then it should work

**Comment by aserrallerios** [comment]
Well, I could do that, and of course it'd work, but that wouldn't fix the issue.

What I think it's happening:

1. My book has the tag `Religion`
2. In my DB exist two tags: `Religion` (899) and `religión` (839)
3. The code thinks that the tag for my book must be `religión` (839) and totally ignores that there's another tag that exists with the label `Religion`
4. While trying to assign the tag 839, it tries to update the label to `Religion`, because even though the code at **3** was smart enough to perform an approximate query, now it deems the label `religión` as not good enough for the tag

The combination of points 3 and 4 produce a bug. Probably 4 should never happen, or 3 should be smarter.

**Comment by OzzieIsaacs** [comment]
@aserrallerios Please check the latest nightly version, now the case with a library already having the "doppelganger" name should be handled properly

**Comment by aserrallerios** [comment]
Sorry, I didn't have the time to test it until now, but it works perfectly :)

**Comment by Oredna** [comment]
I had the same problem. I had two authors - one named with ' and the other with another version of '. I had to remove the author's name with the strange ' and make it unknown. Only then I was able to name it with the original ' and save new books.
